### ISSUE TITLE
index issue to compute KL divergence

### ISSUE THREAD

**Issue by WeiXiongUST** [huggingface/trl#633]
Hi! I am confused about the way how TRL computes the KL divergence, where I quote the implementation as follows for your convenience.

```python
query = ['How are you? Can I have a cup of coffee?', "A coffee will be a good start of the day."
inputs = tokenizer(query, return tensors="pt", padding=True).to(0)
logits = model(**inputs)['logits']
logp = torch.nn.functional.log softmax(logits[:, :-1, :], dim=2)
logprobs = torch.gather(logp, 2, input ids[:, 1:].unsqueeze(2)).squeeze(-1)
```

As we can see, we discard the **last** element in the logits and the **first** element in the input ids. Could you elaborate on this a little bit?

Many thanks!

**Comment by WeiXiongUST** [comment]
oh I get it. The implementation is correct! just ignore it. thanks!
